src/document/utils/base.ts

```diff
@@ -166,6 +175,56 @@ export function readOnly<T>(value: T): Immutable<T> {
     return castImmutable(freeze(value, true));
 }
 
+
+/**
+ * Maps skipped operations in an array of operations.
+ * Skipped operations are operations that are ignored during processing.
+ * @param operations - The array of operations to map.
+ * @returns An array of mapped operations with ignore flag indicating if the operation is skipped.
+ * @throws Error if the operation index is invalid and there are missing operations.
+ */
+export function mapSkippedOperations<A extends Action>(
```

Comment:
this is just a stylistic thing so feel free to ignore if you don't agree,

but I think that since this map function is mutating variables outside its scope and therefore is not pure, it would be better to use explicit iteration instead like:

```ts
export function mapSkippedOperations<A extends Action>(
    operations: Operation<BaseAction | A>[],
): MappedOperation<A>[] {
    const ops = [...operations].reverse();

    let skipped = 0;
    let latestOpIndex = ops.length > 0
        ? ops[ops.length - 1].index
        : 0;

    const scopeOpsWithIgnore = [] as MappedOperation<A>[];

    for (const operation of ops) {

        if (skipped > 0) {
            const operationsDiff = latestOpIndex - operation.index;
            skipped -= operationsDiff
        }
        


        if (skipped < 0) {
            throw new Error('Invalid operation index, missing operations');
        }
        
        const mappedOp = {
            ignore: skipped > 0,
            operation,
        };


        if (operation.skip > 0) {
            // here we add 1 to the skip number because we want to get the number of
            // operations that we want to move the pointer back to get the latest valid operation
            // operation.skip = 1 means that we want to move the pointer back 2 operations to get to the latest valid operation
            skipped = skipped + (operation.skip + 1);
        }

        latestOpIndex = operation.index;
        
        scopeOpsWithIgnore.push(mappedOp);
    }

    return scopeOpsWithIgnore.reverse();
}
```

Reply:
makes sense 👍🏻 
added change here: [3629934](https://github.com/powerhouse-inc/document-model/pull/26/commits/3629934a65d62e2fa3d21dde766dee07d584368d)